src/backend/puppet/observed.rs

```diff
@@ -0,0 +1,585 @@
+    pub fn is_continuation(&self) -> bool {
+        match self {
+            &GraphemePart::Continuation => true,
+            _ => false,
+        }
+    }
+
+    /// Returns Some(String) if GraphemePart is Begin(String), else None.
+    pub fn as_option(&self) -> Option<&String> {
+        match self {
+            &GraphemePart::Begin(ref string) => Some(string),
+            &GraphemePart::Continuation => None,
+        }
+    }
+
+    /// Returns String if GraphemePart is Begin(String), panics otherwise.
+    pub fn unwrap(&self) -> String {
+        match self {
+            &GraphemePart::Begin(ref s) => s.clone(),
+            _ => panic!("unwrapping GraphemePart::Continuation"),
+        }
+    }
+}
+
+/// Represents a single cell of terminal.
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub struct ObservedCell {
+    /// Absolute position
+    pub pos: Vec2,
+    /// Style
+    pub style: Rc<ObservedStyle>,
+    /// Part of grapheme - either it's beginning or continuation when character is multi-cell long.
+    pub letter: GraphemePart,
+}
+
+impl ObservedCell {
+    /// Constructor
+    pub fn new(
+        pos: Vec2,
+        style: Rc<ObservedStyle>,
+        letter: Option<String>,
+    ) -> Self {
+        let letter: GraphemePart = match letter {
+            Some(s) => GraphemePart::Begin(s),
+            None => GraphemePart::Continuation,
+        };
+
+        ObservedCell { pos, style, letter }
+    }
+}
+
+/// Puppet backend output
+///
+/// Represents single frame.
+#[derive(Debug, Clone)]
+pub struct ObservedScreen {
+    /// Size
+    size: Vec2,
+    /// Contents. Each cell can be set or empty.
+    contents: Vec<Option<ObservedCell>>,
+}
+
+impl ObservedScreen {
+    /// Creates empty ObservedScreen
+    pub fn new(size: Vec2) -> Self {
+        let contents: Vec<Option<ObservedCell>> = vec![None; size.x * size.y];
+
+        ObservedScreen { size, contents }
+    }
+
+    fn flatten_index(&self, index: &Vec2) -> usize {
+        assert!(index.x < self.size.x);
+        assert!(index.y < self.size.y);
+
+        index.y * self.size.x + index.x
+    }
+
+    fn unflatten_index(&self, index: usize) -> Vec2 {
+        assert!(index < self.contents.len());
+
+        Vec2::new(index / self.size.x, index % self.size.x)
+    }
+
+    /// Sets all cells to empty cells with given style
+    pub fn clear(&mut self, style: &Rc<ObservedStyle>) {
+        for idx in 0..self.contents.len() {
+            self.contents[idx] = Some(ObservedCell::new(
+                self.unflatten_index(idx),
+                style.clone(),
+                None,
+            ))
+        }
+    }
+
+    /// Size
+    pub fn size(&self) -> Vec2 {
+        self.size
+    }
+
+    /// Returns a rectangular subset of observed screen.
+    pub fn piece(&self, min: Vec2, max: Vec2) -> ObservedPiece {
+        ObservedPiece::new(self, min, max)
+    }
+
+
+    /// Returns occurences of given string pattern
+    pub fn find_occurences<'a>(
```

Comment:
Could this be implemented for `ObservedPieceInterface`? It seems to use `self.min()` and `self.max()`.

Reply:
you are probably right, but for some reason it does not allow me to use index operator on `ObservedPieceInterface`.